### ISSUE TITLE
[BUG] Test Reporter Gets Confused on File Name

### ISSUE THREAD

**Issue by rwoll** [microsoft/playwright#11126]
**Context:**
- Playwright Version: 1.17 and ToT
- Operating System: macOS
 - Node: 14.17.6 - ~/.nvm/versions/node/v14.17.6/bin/node
 - Yarn: 1.22.11 - ~/.nvm/versions/node/v14.17.6/bin/yarn
 - npm: 7.24.1 - ~/.nvm/versions/node/v14.17.6/bin/npm

**Code Snippet**

See the following for a full description and repro: https://github.com/rwoll/playwright-parameterized-tests-repro/blob/main/README.md 

**Describe the bug**

We define a test parametizer/helper `parameterizeTestCase` and call it in various other test files to get shared behavior.

(In reality the actual `parameterizeTestCase` has a much more extensive implementation to get templated test cases for API testing.)

If you run this sample, you will see the following output:

## Actual

```
$ npm run test:e2e

> plawyright-parameterized-tests-repro@1.0.0 test:e2e
> playwright test


Running 3 tests using 3 workers

  ✓  helpers/index.ts:4:3 › a (6ms)
  ✓  helpers/index.ts:4:3 › suite › b-0 (5ms)
[ 'e2e/a.spec.ts', 'a' ]
  ✓  helpers/index.ts:4:3 › suite › b-1 (5ms)
[ 'e2e/b.spec.ts', 'suite', 'b-0' ]
[ 'e2e/b.spec.ts', 'suite', 'b-1' ]
```

## Expected

```
$ npm run test:e2e

> plawyright-parameterized-tests-repro@1.0.0 test:e2e
> playwright test


Running 3 tests using 3 workers

  ✓  e2e/a.spec.ts:?:? › a (6ms)
  ✓  e2e/b.spec.ts:?:? › suite › b-0 (5ms)
[ 'e2e/a.spec.ts', 'a' ]
  ✓  e2e/b.spec.ts:?:? › suite › b-1 (5ms)
[ 'e2e/b.spec.ts', 'suite', 'b-0' ]
[ 'e2e/b.spec.ts', 'suite', 'b-1' ]
```

I expected the reporter to not reference `helpers/index.ts` at all.

A few notes:
* The `testInfo.titlePath` looks correct in the actual output.
* It looks like the runner chose to run `a.spec.ts` and `b.spec.ts` in parallel (based on worker counts).
* HTML Report works correctly, and doesn't have the same reporting issue. (See it via `npx playwright test --reporter html && npx playwright show-report`).

**Comment by rwoll** [comment]
Correction: The HTML Reporter gets it partially correct. It groups the tests properly so they are in the correct accordion, but the line items within them still reference `helpers/index.ts` within the case. For example, see this screenshot:

<img width="1056" alt="Screen Shot 2021-12-28 at 7 28 10 PM" src="https://user-images.githubusercontent.com/11915034/147624720-04f59050-ca47-4375-93e5-c2947796ed25.png">

**Comment by aslushnikov** [comment]
@dgozman I wonder if we can prefer `.spec.ts` or `.spec.js` frames to the top stack frame when determining test location?

**Comment by dfberry** [comment]
I have the same problem.
OS: mac
Nodejs: 14.15.1
NPM 6.14.8

The error is in the authentication spec but is only partially reported that way. 

```
> e2e@1.0.0 test /Users/dfberry/repos/github-who/e2e
> npx playwright test --reporter=list

Using config at /Users/dfberry/repos/github-who/e2e/playwright.config.ts

Running 12 tests using 2 workers

  ✓  [chromium] › basic.failing.spec.ts:3:1 › basic not found test (3s)
  ✓  [chromium] › authentication.spec.ts:4:1 › Authenticate, get profile (1s)
  ✓  [chromium] › basic.spec.ts:3:1 › basic found test (2s)
  ✓  [chromium] › example.spec.ts:3:1 › basic test (4s)
  ✓  [firefox] › authentication.spec.ts:4:1 › Authenticate, get profile (4s)
  ✓  [firefox] › basic.failing.spec.ts:3:1 › basic not found test (10s)
  ✓  [firefox] › basic.spec.ts:3:1 › basic found test (4s)
  ✘  [firefox] › example.spec.ts:3:1 › basic test (5s)
  ✓  [webkit] › authentication.spec.ts:4:1 › Authenticate, get profile (2s)
  ✓  [webkit] › basic.failing.spec.ts:3:1 › basic not found test (3s)
  ✓  [webkit] › basic.spec.ts:3:1 › basic found test (3s)
  ✓  [webkit] › example.spec.ts:3:1 › basic test (7s)


  1) [firefox] › example.spec.ts:3:1 › basic test ==================================================

    Error: expect(received).toEqual(expected) // deep equality

    Expected: "george ezra"
    Received: {"_frame": {"_guid": "frame@a040dfa86bb18ee06201d7b1b53dc9ec", "_type": "Frame"}, "_selector": ".GitHubProfile"}

        at /Users/dfberry/repos/github-who/e2e/tests/authentication.spec.ts:37:25

    attachment #1: trace (application/zip) ---------------------------------------------------------
    test-results/example-basic-test-firefox/trace.zip
    Usage:

        npx playwright show-trace test-results/example-basic-test-firefox/trace.zip

    ------------------------------------------------------------------------------------------------


  1 failed
    [firefox] › example.spec.ts:3:1 › basic test ===================================================
  11 passed (33s)
```
![Screen Shot 2022-01-01 at 9 54 09 AM](https://user-images.githubusercontent.com/1517008/147857047-3a593049-29a9-4139-a561-813857f89528.png)

**Comment by dgozman** [comment]
@dfberry Could you please share more details? I don't quite understand your setup, for example how is "authentication.spec" ended up being called from the "example.spec"?

@rwoll There are multiple scenarios to consider, and we should find the right balance between them.

1. `const myTestWrapper = require('./helper')` like in your scenario, that is called like `myTestWrapper()` from the `example.spec.ts` and generates a single test. Ideally, this is reported as **example.spec.ts:5:1**.
2. `require('./my-test-3'); require('./my-test-4')` directly in `example.spec.ts`, where both files call `test` inside. Ideally, this is reported as **my-test-3.ts:5:1**.
3. `const testInMultipleConfigurations = require('./helper')`, that is called like `testInMultipleConfigurations([1, 2, 3])` from the `example.spec.ts` and generates three tests. Ideally, this is reported as **helper.ts:5:3**, because there are multiple tests from a single call in example.spec?

I have hard time distinguishing between cases 1 and 3, although the expectations are most likely different.

**Comment by rwoll** [comment]
> @dfberry Could you please share more details? I don't quite understand your setup, for example how is "authentication.spec" ended up being called from the "example.spec"?
> 
> @rwoll There are multiple scenarios to consider, and we should find the right balance between them.
> 
> 1. `const myTestWrapper = require('./helper')` like in your scenario, that is called like `myTestWrapper()` from the `example.spec.ts` and generates a single test. Ideally, this is reported as **example.spec.ts:5:1**.
> 2. `require('./my-test-3'); require('./my-test-4')` directly in `example.spec.ts`, where both files call `test` inside. Ideally, this is reported as **my-test-3.ts:5:1**.
> 3. `const testInMultipleConfigurations = require('./helper')`, that is called like `testInMultipleConfigurations([1, 2, 3])` from the `example.spec.ts` and generates three tests. Ideally, this is reported as **helper.ts:5:3**, because there are multiple tests from a single call in example.spec?
> 
> I have hard time distinguishing between cases 1 and 3, although the expectations are most likely different.

@dgozman Can you show some examples of what are in the `helper` for (1) and (3) for your above examples? In my mind they are logically equivalent and should report as being apart of `example.spec.ts`. ((3) is just (1) with a for loop.)

My repro repo had an extremely abbreviated example, in reality our `parameterizeTestCase` looks something more like[^1]:

```ts
export const parameterizedTestCase = (args: {
  description: string;
  focus?: boolean;
  target: string;
  enabledAnalysisEngines: string[];
  apiOptions?: Record<string, string>;
  additionalAssertions?: (job: Job) => Promise<void>;
}) => {
  (args.focus ? test.describe.serial.only : test.describe.serial)(
    `API Snaphost Test: ${args.description}`,
    () => {
      test.skip(!CONFIG.enginesUnderTest.has(args.enabledAnalysisEngines), "Engine needed for test is not enabled.");

      let job: Job | null = null;

      test("submit", async ({ submit }) => {
        const rawResponse = submit(
          args.target,
          args.enabledAnalysisEngines,
          args.apiOptions
        );
        // validateAPIResponse internalluy asserts/`expect` status codes and validates schema; throws errors via expect along the way for any issues
        job = validateAPIResponse(rawResponse);
      });

      test("snapshot", async ({}, testInfo) => {
        expect(job).not.toBeNull();
        const j: Job = job!;
        expect(normalize(j).toYAML({ sort: true })).toMatchSnapshot(
          `${sanitizeFileName(testInfo.titlePath.join("-"))}.yml`
        );
      });

      test("custom assertions", async () => {
        expect(job).not.toBeNull();
        const j: Job = job!;
        // additionalAssertions will have expect calls in it for some test cases
        if (args.additionalAssertions) await args.additionalAssertions(j);
      });
    }
  );
};
```

I could see us adding another layer to manage API options, and run the same set of tests with different API options, but I'd still expect it to report under `example.spec.ts` and never where the helper is defined.

`*.spec.ts` files we use to organize the tests into logical groups even if their implementation (i.e. the helpers they each call are all the same).

Our previous e2e API+smoketest suite implemented via pytest had 100ish tests that simply extended a base test class and overrode an `additional_assertions` method. 

[^1]: It looks a bit different in reality in implementation, typing, etc. but hopefully this gives the gist without requiring too much context.

**Comment by rwoll** [comment]
@dgozman Aside: once we have a more concrete idea of how each of these should report/group, I'd like to follow up on the report info, code frames, and stack traces. For example, if the `expect` is in a helper that's called in a test:

```
test('…', () => {
    …
    helpers.commonAsserts(…)
})
```

how much code frame context should be shown in the reporter? Should it show `helpers.commonAsserts` callsite as the problem code frame, or should it dive into the `helpers.commonAsserts` where expect was called? Perhaps, the answer is one of (1) user should own expect extensions instead of using a helper/wrapper, (2) in the displayed stack trace, make the stack trace clickable so you can dive further in/out to the files/lines mentioned in the currently displayed trace, (3) show more frames (e.g. one per line in the trace) but only in userland, not all the way into node_modules/core node, or (4) some combination of these :)

Let me know if this aside/issue/problem makes sense and whether we want that discussion to happen on this ticket (after we sort out behavior around which filename the test cases should report under), or if you'd like me to follow up with a separate ticket.

Thanks!

**Comment by dgozman** [comment]
@rwoll We usually suggest to use helper functions instead of test wrappers:

```js
async function helper(submit: SubmitFixtureType, testInfo: TestInfo, args: {
  target: string;
  apiOptions?: Record<string, string>;
  enabledAnalysisEngines: string[];
  additionalAssertions?: (job: Job) => Promise<void>;
}) {
  test.skip(!CONFIG.enginesUnderTest.has(enabledAnalysisEngines), "Engine needed for test is not enabled.");

  let job: Job;

  await test.step('submit', async () => {
    const rawResponse = submit(
      args.target,
      args.enabledAnalysisEngines,
      args.apiOptions
    );
    // validateAPIResponse internalluy asserts/`expect` status codes and validates schema; throws errors via expect along the way for any issues
    job = validateAPIResponse(rawResponse);
  });

  await test.step('snapshot', async () => {
    expect(job).not.toBeNull();
    expect(normalize(job).toYAML({ sort: true })).toMatchSnapshot(
      `${sanitizeFileName(testInfo.titlePath.join("-"))}.yml`
    );
  });

  await test.step('custom assertions', async () => {
    // additionalAssertions will have expect calls in it for some test cases
    if (args.additionalAssertions) await args.additionalAssertions(job);
  });
}

// Now use the helper as many times as we want.
test(`API Snaphost Test: description one`, async ({ submit }, testInfo) => {
  await helper(submit, testInfo, {
    target: 'target',
    apiOptions: {},
    enabledAnalysisEngines: ['foo', 'bar'],
  });
});
```

This way you are not going against the test runner, but work together with it.

Re: the error snippet - I will look into that.

**Comment by rwoll** [comment]
Thanks, @dgozman! 

I'll switch to the helper with explicit test pattern across the board. I don't think that should be an issue here. (We use it elsewhere plus it gives better focus etc. use and probably resolves the "file name confusion issue").

Aside: The actual implementation did use `step`, too, I just paraphrased the code badly when I paraphrased/sanitized it for this ticket but that's actually important since IIRC you can't nest serial and parallel suites, so we need the sub parts of the parameterized test to be steps so we can parallelize the generated tests like:

```
test.describe.parallel("API Tests", () => {
    for (const params of specs)
       test(…, … => helpers…)
});

**Comment by rwoll** [comment]
> @rwoll We usually suggest to use helper functions instead of test wrappers

@dgozman I switched everything over to helper pattern, but one major downside of not having the original setup where the top-level test is collocated with the steps in the helper/parametizer, is every time I add a fixture that needs to be accessed in the helper/steps, it has to be plumbed through at every call site (in this case over 100). Do you have a suggestion around this?

```
// before: doesn't need to be updated if base generatedTest impl changes/improves/etc.
generateTest(params)

// now: this has to be updated every time a fixture is added across all files/tests if helper needs the fixture in it's implementation
test('example', async ({ a, b }, testInfo) => {
  await helpers.execute(params, testInfo, { a, b })
});
```

Thanks!

**Comment by dgozman** [comment]
@rwoll In this case, I'd suggest to make `helper` a fixture itself. This way, it can depend on fixtures it needs directly. For example, in our own tests we have ChildProcessHelper, CLIMock helper, etc.

**Comment by rwoll** [comment]
@dgozman Brilliant! Thank you! (I don't know why I thought I couldn't call `test.step` in a fixture helper/closure, but it of course works 😃 .)
